Fix `proc macro panicked` with `Invalid type` in `const_random!` in Rust 1.77.

nnethercote opened the pull request:
This PR fixes a compile-time panic with `const_random!` on byte arrays in Rust 1.77. Full details are in the individual commit logs.

dtolnay reviewed (changes requested):
I believe the `usize`/`isize` change is not correct, because it assumes the host's pointer size is at least as large as the target's pointer size. If usize is 4 bytes on the host and 8 bytes on the target, const_random would be producing random usizes that are half zero.

nnethercote commented:
Well spotted, thanks. I have updated to remove the first commit changing `usize`/`isize` handling. I also added a comment explaining the subtlety, so nobody else makes the same mistake in the future.

dtolnay commented on macro/src/lib.rs:
```diff
@@ -14,44 +12,89 @@ fn ident(ident: &str) -> TokenStream {
+                    // `usize::random()` would produce a 32-bit random usize which would then be
+                    // turned into a suffixed literal (e.g. `0x1234_5678usize`). On the 64-bit
+                    // target that literal would always have the upper 32 bits as zero, which would
+                    // be bad. Instead we produce code that will generate a 128-bit integer literal
+                    // (on the target) and then truncate it to usize (on the host).
```
```suggestion
                    // be bad. Instead we produce code that will generate a 128-bit integer literal
                    // (on the host) and then truncate it to usize (on the target).
```

nnethercote replied:
Fixed, thanks.

dtolnay reviewed (approved):
Thank you! This looks good to me.

nnethercote commented:
Oh, this problem may also occur in Rust 1.76.

decathorpe commented:
Would it be possible to get a new version of const-random-macros released with this fix?
